bench/benches/operators.rs

```diff
@@ -0,0 +1,178 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+use nb2::batch::Either;
+use nb2::packets::{Ethernet, Packet};
+use nb2::testils::criterion::BencherExt;
+use nb2::testils::proptest::*;
+use nb2::{compose, Batch, Mbuf};
+use proptest::prelude::*;
+use proptest::strategy;
+
+const BATCH_SIZE: usize = 500;
+static mut COUNTER: u32 = 0;
+
+fn filter_true(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    batch.filter(|_p| true)
+}
+
+fn filter_false(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    batch.filter(|_p| false)
+}
+
+fn filter_map(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Ethernet> {
+    batch.filter_map(|p| {
+        let ethernet = p.parse::<Ethernet>()?;
+        Ok(Either::Keep(ethernet))
+    })
+}
+
+fn map_then_filter(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Ethernet> {
+    batch.map(|p| p.parse::<Ethernet>()).filter(|_p| true)
+}
+
+fn map(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Ethernet> {
+    batch.map(|p| p.parse::<Ethernet>())
+}
+
+fn group_by(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    unsafe { COUNTER += 1 };
+
+    unsafe {
+        batch.group_by(
+            |_p| COUNTER,
+            |groups| {
+                compose!(groups {
+                    0 => |group| {
+                        group.for_each(|_p| Ok(()))
+                    }
+                    _ => |group| {
+                        group.for_each(|_p| Ok(()))
+                    }
+                })
+            },
+        )
+    }
+}
+
+fn replace(batch: impl Batch<Item = Mbuf>) -> impl Batch<Item = Mbuf> {
+    batch.replace(|_p| Mbuf::new())
+}
+
+fn no_batch_replace(_mbuf: Mbuf) -> Mbuf {
+    Mbuf::new().unwrap()
+}
+
+#[nb2::bench(mempool_capacity = 511)]
+fn filters_batch(c: &mut Criterion) {
```

Comment:
very minor, maybe move this up right next to the routines being grouped? I scroll up and down quite a bit to look at them.

Reply:
So, fn then bench? Should packets work that way too?

Reply:
I think that's a bit easier to read. but that's just my preference cuz I don't know how to navigate source code file other than scrolling up and down. 😄